#! /bin/sh /usr/share/dpatch/dpatch-run
## 06_support_fromiso_isofrom.dpatch <mika@grml.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Support isofrom / fromiso bootoption

@DPATCH@

--- a/scripts/live
+++ b/scripts/live
@@ -92,6 +92,11 @@ Arguments ()
 				export HOSTNAME LIVECONF
 				;;
 
+			isofrom=*|fromiso=*)
+				FROMISO="${ARGUMENT#*=}"
+				export FROMISO
+				;;
+
 			username=*)
 				USERNAME="${ARGUMENT#username=}"
 				LIVECONF="changed"
@@ -1361,6 +1366,17 @@ check_dev ()
 	devname="${2}"
 	skip_uuid_check="${3}"
 
+	# support for fromiso=.../isofrom=....
+	if [ -n "$FROMISO" ]
+	then
+		mkdir /isofrom
+		ISO_DEVICE="$(echo $FROMISO | sed 's|\(/dev/[a-z]*[0-9]*\).*|\1|')"
+		mount "$ISO_DEVICE" /isofrom
+		ISO_NAME="$(echo $FROMISO | sed 's|/dev/[a-z]*[0-9]*/||')"
+		loopdevname=$(setup_loop "/isofrom/${ISO_NAME}" "loop" "/sys/block/loop*" "" '')
+		devname="${loopdevname}"
+	fi
+
 	if [ -z "${devname}" ]
 	then
 		devname=$(sys2dev "${sysdev}")
@@ -1632,6 +1648,15 @@ mountroot ()
 		log_end_msg
 	fi
 
+	# if we do not unmount the ISO we can't run "fsck /dev/ice" later on
+	# because the mountpoint is left behind in /proc/mounts, so let's get
+	# rid of it when running from RAM
+	if [ -n "$FROMISO" ] && [ "${TORAM}" ]
+	then
+	  losetup -d /dev/loop0
+	  grep -q /isofrom /proc/mounts && umount /isofrom
+	fi
+
 	if [ -n "${MODULETORAMFILE}" ] || [ -n "${PLAIN_ROOT}" ]
 	then
 		setup_unionfs "${livefs_root}" "${rootmnt}"
